### PR TITLE
MemoryExec INSERT INTO refactor to use ExecutionPlan

### LINKED ISSUE #6008: Implement INSERT INTO as part of LogicalPlan and PhysicalPlan
### Is your feature request related to a problem or challenge?

INSERT INTO is currently implemented separately from the query planner, which is not an ideal approach. To improve the design and align with major database engines, we propose to add support for inserting data into LogicalPlan and PhysicalPlan.

**Current Behavior:**

INSERT INTO operation is handled separately from the query planner.

**Desired Behavior:**

INSERT INTO operation should be integrated into LogicalPlan and PhysicalPlan.

**Benefits:**

Aligns with the design of major database engines.
Simplifies the codebase by handling all operations within the query planner.
Potentially improves performance by enabling more query optimization opportunities.

### Describe the solution you'd like

INSERT INTO operation should be integrated into LogicalPlan and PhysicalPlan.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by metesynnada**
# Which issue does this PR close?

Closes #6008.

# Rationale for this change

INSERT INTO is currently implemented separately from the query planner, which is avoidable. Like major engines, we can add inserting into `LogicalPlan` and `PhysicalPlan`.

# What changes are included in this PR?
- How a MemTable holds its `RecordBatch` data,
- Moving insert dml logical plan into into `create_initial_plan` instead of separate plan addition,
- `MemoryWriterExec` for execution.

# Are these changes tested?
Yes

# Are there any user-facing changes?

Now, to execute the insert query, the user needs to call collect, as in all major engines.

### CODE REVIEW

**Review by mustafasrepo — APPROVED** [review]
This PR is passed from out interval review process. It is LGTM!.

**Review by alamb — COMMENTED** [review]
Thank you -- I would like to review this PR more closely before merge. 

I mostly want to ensure the boundary in DataFusion between "query" and "database system" remains clear. Among other things I want to ensure that people can use DataFusion as a read only query engine as well as to build more full featured DBMS systems

**Review by alamb — COMMENTED** [review]
Thank you @metesynnada

I think the change to `TableProvider::insert_into` to take `ExecutionPlan` is reasonable  one as it allows the source partitioning to be maintained. 

However, I don't understand the value of creating an `ExecutionPlan` as output -- since the resulting `ExecutionPlan`  has no output (it routes the inputs internally) most of the method on `ExecutionPlan` don't make sense (like "output ordering", and "preserves input") 

Thus I think this PR could be significantly simplified without loss of functionality

What do you think @metesynnada ?

**Comment by metesynnada** [comment]
Thank you for reviewing and providing valuable feedback. I appreciate your input on the change to `TableProvider::insert_into` to accept an `ExecutionPlan`. My intention behind this implementation was to align with the discussion in the #5076, which suggests re-implementing DataFrame.write_* methods to use a LogicalPlan::Write operation along with a physical operator to perform the write.

The rationale behind this approach is to create a more unified and extensible architecture for handling write operations, similar to other major engines like Apache Spark, which incorporates write operations in both logical and physical plans.

I understand your concerns about creating an `ExecutionPlan` as output, specifically regarding the methods that might not make sense in this context, like "output ordering" and "preserves input." In light of the related issue, I believe that integrating write operations into the `ExecutionPlan` can help address the need for a more generic `LogicalPlan::Write ` operation. `INSERT INTO` support was just the beginning.

If you believe that simplifying the PR without losing functionality is the best course of action, I am willing to make the necessary changes. However, I would appreciate further discussion on how to align this PR with the goals outlined in issue #5076 while addressing your concerns about the resulting `ExecutionPlan`.

Thank you again for your feedback, and I look forward to refining this PR with your assistance and in accordance with the project's roadmap.

**Comment by alamb** [comment]
So I guess I would prefer one of the following
1. Remove the ExecutionPlan from this PR
2. Make the ExecutionPlan more general (work for any table provider) rather than just the MemExec

File: datafusion/core/src/physical_plan/memory_insert.rs
Comment (tustvold) [inline datafusion/core/src/physical_plan/memory_insert.rs:201]: FWIW I think this could be implemented more concisely using [`futures::unfold`](https://docs.rs/futures/latest/futures/prelude/stream/fn.unfold.html) and [`RecordBatchStreamAdapter`](https://docs.rs/datafusion/latest/datafusion/physical_plan/stream/struct.RecordBatchStreamAdapter.html).

File: datafusion/core/src/physical_plan/memory_insert.rs
Comment (tustvold) [inline datafusion/core/src/physical_plan/memory_insert.rs:135]: FWIW I think this could be implemented more concisely using [`futures::unfold`](https://docs.rs/futures/latest/futures/prelude/stream/fn.unfold.html) and [`RecordBatchStreamAdapter`](https://docs.rs/datafusion/latest/datafusion/physical_plan/stream/struct.RecordBatchStreamAdapter.html).

File: datafusion/core/src/physical_plan/memory_insert.rs
Comment (tustvold) [inline datafusion/core/src/physical_plan/memory_insert.rs:45]: The Box is redundant if calling poll_unpin

File: datafusion/core/src/physical_plan/memory_insert.rs
Comment (tustvold) [inline datafusion/core/src/physical_plan/memory_insert.rs:54]: More an observation, but relates to https://github.com/apache/arrow-datafusion/issues/5076, it seems like MemoryWriteExec is very tightly coupled with the internal data structures of MemTable. I think this is inevitable for all `TableProvider`, which makes me wonder if this should really just be an internal implementation detail of a given `TableProvider`, as it will be highly specific?

Comment (alamb) [reply datafusion/core/src/physical_plan/memory_insert.rs:54]: > which makes me wonder if this should really just be an internal implementation detail of a given TableProvider, as it will be highly specific?

yes, this is my preference as well (mentioned in https://github.com/apache/arrow-datafusion/pull/6049#issuecomment-1520524109)

File: datafusion/core/src/datasource/memory.rs
Comment (metesynnada) [reply datafusion/core/src/datasource/memory.rs:46]: You are right, we can move on with `Vec<Arc<RwLock<Vec<RecordBatch>>>>`.

File: datafusion/core/src/physical_plan/memory_insert.rs
Comment (metesynnada) [reply datafusion/core/src/physical_plan/memory_insert.rs:135]: I am not sure how to acquire the async lock within the fold, for once. Also, what you are folding onto will be the output, it is unclear to me how to make it result `Result<RecordBatch>`. Can you suggest it here?

File: datafusion/core/src/physical_plan/memory_insert.rs
Comment (metesynnada) [reply datafusion/core/src/physical_plan/memory_insert.rs:45]: Did you try it in code? It will produce an error since lock future (`impl Future<Output = RwLockWriteGuard<'_, Vec<RecordBatch>>>`) does not implement `Unpin`. `BoxFuture`s are implementing this, I think you made the same thing in `FileStreamState`?

Is there another poll method that does not require pinning the future? I could use that.

**Comment by metesynnada** [comment]
Thanks for the review @alamb and @tustvold , just sent a commit to address your concerns. We moved `MemoryWriteExec` to `memory.rs` and removed the new module in the physical plan. Now, `memory.rs` contains the operator reading from it (`MemoryExec`) and the operator writing into it (`MemoryWriteExec`). We also made some fixes per Rafael’s comments. Can you please take a look and share your thoughts? Thanks.

File: datafusion/core/src/datasource/datasource.rs
Comment (aprimadi) [reply datafusion/core/src/datasource/datasource.rs:106]: Just a question because I'm kinda new to this project. If `insert_into` takes as input `Vec<SendableRecordBatchStream>`, shouldn't it be able to run multiple partitions in parallel?

Comment (metesynnada) [reply datafusion/core/src/datasource/datasource.rs:106]: The current implementation is already executing the writing parallel. The proposed way is changing where it is handled.

File: datafusion/core/src/physical_plan/memory.rs
Comment (ozankabak) [reply datafusion/core/src/physical_plan/memory.rs:249]: > But I thought the consensus was to have a generic ExecutionPlan for writing data, rather than one that was specific to MemTable

Are you sure about this? My understanding was going through the `ExecutionPlan` route for each writer, not a single plan for everything. There is a separate `Exec` on the read side for every `TableProvider`, why would we not follow the same pattern?

Anyways, I asked @metesynnada to experiment with the single plan idea, but he wasn't able to make progress on it -- he will comment here with his findings.

Comment (tustvold) [reply datafusion/core/src/physical_plan/memory.rs:249]: I agree that a single plan is likely not tractable, it will be highly specific to the `TableProvider` in question. What I had meant was simply not returning anything and leaving everything an internal implementation detail of the `TableProvider` from how it executes the provided `ExecutionPlan` through to how that is exposed. Not only would this make this PR significantly simpler, obviating the need for any of the ExecutionPlan or Stream shenanigans, but would provide the most flexibility for table implementations that have more complex scheduling or transaction handling requirements. Exposing `ExecutionPlan` is a pretty strong API commitment, and I _personally_ am not confident this won't be painting ourselves into a corner, but if the consensus is otherwise :shrug:

Comment (alamb) [reply datafusion/core/src/physical_plan/memory.rs:249]: > All in all, having a distinct ExecutionPlan to handle writes for each data source mirrors the behavior on the read side and seems much more natural. What do you think?

I guess I don't fully understand why adding 800 lines of `MemTable` specific ExecutionPlan  that seems to add no new functionality is an improvement

Not only only this this  pattern seem to  overly complicate this PR, but it sets us on a course for every other table source to have to do the same (mostly repeated) things.

Perhaps there is a future use case I don't understand.

If you guys feel strongly that this is the right and maintainable solution going forward I will defer to your judgement, but from where I see it just adds complexity / technical debt for no reason.

Comment (ozankabak) [reply datafusion/core/src/physical_plan/memory.rs:249]: This is simply a baby step for implementing things like `COPY TO`s, `INSERT INTO`s and functionality like that. In practical cases, these things will write their outputs through a `TableProvider` like a `ListingTable`. How do we test what that will look like and iterate on such functionality? Basically, this is the simplest writer that one can perform end-to-end (read-transform-write) experiments with, its main intention is for anyone to use as a basis/example for implementing their write operators. I hope it makes a little more sense now.

> If you guys feel strongly that this is the right and maintainable solution going forward I will defer to your judgement, but from where I see it just adds complexity / technical debt for no reason.

Thank you -- rest assured we will be maintaining and helping with related work going forward.

**Review by alamb — APPROVED** [review]
I vote we merge this PR and keep iterating in subsequent PRs

Thank you for the discussions @metesynnada  and @ozankabak and @tustvold

File: datafusion/core/src/physical_plan/memory.rs
Comment (tustvold) [inline datafusion/core/src/physical_plan/memory.rs:400]: Does this need to be public?

File: datafusion/core/src/physical_plan/memory.rs
Comment (tustvold) [inline datafusion/core/src/physical_plan/memory.rs:503]: Same here, I would keep these private

**Comment by tustvold** [comment]
I've filed https://github.com/apache/arrow-datafusion/pull/6154 to try to refactor the code to be a little clearer, PTAL

**Pull request merged**
